### PR TITLE
CVSB-11333 new function added for updating EuVehicleCategory with code coverage

### PR DESCRIPTION
**Pull request by m-ujaffer**
https://jira.dvsacloud.uk/browse/CVSB-11333

### CODE REVIEW

File: src/functions/updateEuVehicleCategory.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+export async function updateEuVehicleCategory(event: any) {
+    const s3BucketService = new S3BucketService(new S3());
+    const techRecordsService = new TechRecordsService(new TechRecordsDAO(), s3BucketService);
+
+    const vin: string = event.pathParameters!.vin;
+    const euVehicleCategoryString: string = event.queryStringParameters.euVehicleCategory;
+    const euVehicleCategory = EU_VEHICLE_CATEGORY[euVehicleCategoryString.toUpperCase() as keyof typeof EU_VEHICLE_CATEGORY];
+    try {
+        const response = await techRecordsService.updateEuVehicleCategory(vin, euVehicleCategory);
```
Comment (KHodderDVSA) [inline src/functions/updateEuVehicleCategory.ts]: can do this in a single line

File: src/services/TechRecordsService.ts [hunk 2]
```diff
@@ -331,6 +332,32 @@ class TechRecordsService {
...
     return notifiableAlterationIds.includes(testTypeId);
   }
+
+  public async updateEuVehicleCategory(vin: string, newEuVehicleCategory: EU_VEHICLE_CATEGORY): Promise<HTTPResponse | HTTPError> {
+    const techRecordWrapper: ITechRecordWrapper = await this.getTechRecordsList(vin, STATUS.ALL);
+    const nonArchivedTechRecord = techRecordWrapper.techRecord.filter((techRecord) => techRecord.statusCode !== STATUS.ARCHIVED);
+    if (nonArchivedTechRecord.length === 1) {
+      if(nonArchivedTechRecord[0].euVehicleCategory) {
+        return new HTTPResponse(200,HTTPRESPONSE.NO_EU_VEHCILE_CATEGORY_UPDATE_REQUIRED);
+      } else {
```
Comment (KHodderDVSA) [inline src/services/TechRecordsService.ts]: don't need the else. Return will escape by itself.

File: src/services/TechRecordsService.ts [hunk 3]
```diff
@@ -331,6 +332,32 @@ class TechRecordsService {
...
+        let updatedTechRecord;
+        try {
+          updatedTechRecord =  await this.techRecordsDAO.updateSingle(techRecordWrapper);
+        } catch (error) {
+          throw new HTTPError(500, HTTPRESPONSE.INTERNAL_SERVER_ERROR);
+        }
+        return new HTTPResponse(200, this.formatTechRecordItemForResponse(updatedTechRecord.Attributes as ITechRecordWrapper));
+      }
+    } else {
+      throw new HTTPError(400, "The vehicle has more than one non archived Tech record!");
```
Comment (KHodderDVSA) [inline src/services/TechRecordsService.ts]: Could reverse the top level If statement, and throw this if length <> 1, in order to reduce nesting/complexity.

File: serverless.yml [hunk 4]
```diff
@@ -135,22 +135,22 @@ functions:
...
             parameters:
               paths:
-                vin: true
+                systemNumber: true
   updateEuVehicleCategory:
     handler: src/handler.handler
     events:
       - http:
-          path: vehicles/{vin}/update-eu-vehicle-category
+          path: vehicles/update-eu-vehicle-category/{systemNumber}
```
Comment (KHodderDVSA) [inline serverless.yml]: this structure change does not match your Update-Status service change

File: src/config/config.yml [hunk 5]
```diff
@@ -18,7 +18,7 @@ functions:
     function: updateTechRecordStatus
 - updateEuVehicleCategory:
     method: PUT
-    path: /vehicles/:vin/update-eu-vehicle-category
+    path: /vehicles/update-eu-vehicle-category/:systemNumber
```
Comment (KHodderDVSA) [inline src/config/config.yml]: and here

File: src/assets/Enums.ts [hunk 6]
```diff
@@ -11,7 +11,11 @@ export enum HTTPRESPONSE {
     RESOURCE_NOT_FOUND = "No resources match the search criteria.",
     INTERNAL_SERVER_ERROR = "Internal Server Error",
     MORE_THAN_ONE_MATCH = "The provided partial VIN returned more than one match.",
-    NO_STATUS_UPDATE_REQUIRED = "No status update required"
+    S3_ERROR = "Upload on S3 failed",
+    S3_DOWNLOAD_ERROR = "Cannot download document from S3",
```
Comment (agoncearenco) [inline src/assets/Enums.ts]: I see that you removed the code related to S3 after rebasing with develop. Shouldn't you remove these also?

File: src/config/config.yml [hunk 7]
```diff
@@ -16,6 +16,15 @@ functions:
...
     function: updateTechRecordStatus
+- updateEuVehicleCategory:
+    method: PUT
+    path: /vehicles/update-eu-vehicle-category/:systemNumber
+    function: updateEuVehicleCategory
+s3:
+  local:
+    endpoint: http://localhost:7000
+    s3ForcePathStyle: true
+  remote: {}
```
Comment (agoncearenco) [inline src/config/config.yml]: I think this S3 config should be removed also

File: serverless.yml (hunk 4 above)
Comment (KHodderDVSA) [reply serverless.yml]: update-status changed to match

File: src/config/config.yml (hunk 5 above)
Comment (KHodderDVSA) [reply src/config/config.yml]: update-status changed to match

File: src/services/TechRecordsService.ts [hunk 8]
```diff
@@ -334,6 +336,30 @@ class TechRecordsService {
...
   }
+
+  public async updateEuVehicleCategory(systemNumber: string, newEuVehicleCategory: EU_VEHICLE_CATEGORY): Promise<HTTPResponse | HTTPError> {
+    const techRecordWrapper: ITechRecordWrapper = (await this.getTechRecordsList(systemNumber, STATUS.ALL, SEARCHCRITERIA.SYSTEM_NUMBER))[0];
+    const nonArchivedTechRecord = techRecordWrapper.techRecord.filter((techRecord) => techRecord.statusCode !== STATUS.ARCHIVED);
+    if (nonArchivedTechRecord.length > 1) {
+        throw new HTTPError(400, HTTPRESPONSE.EU_VEHICLE_CATEGORY_MORE_THAN_ONE_TECH_RECORD);
+    }
+    if(nonArchivedTechRecord[0].euVehicleCategory) {
+        return new HTTPResponse(200,HTTPRESPONSE.NO_EU_VEHCILE_CATEGORY_UPDATE_REQUIRED);
```
Comment (KHodderDVSA) [inline src/services/TechRecordsService.ts]: typo in NO_EU_**VEHCILE**_CATEGORY_UPDATE_REQUIRED

File: tests/unit/handler.unitTest.ts [hunk 9]
```diff
@@ -161,6 +161,31 @@ describe("The lambda function handler", () => {
         expect(TechRecordsService.prototype.updateTechRecordStatusCode).toHaveBeenCalled();
       });
 
+      it("should call the update-eu-vehicle-category function with correct event payload", async () => {
```
Comment (KHodderDVSA) [inline tests/unit/handler.unitTest.ts]: test doesn't actually check the payload

Comment (KHodderDVSA) [reply tests/unit/handler.unitTest.ts]: if you name the mock before assigning it to the prototype, you can use mockName.mock.calls[0].args to get the calling payload

File: src/services/TechRecordsService.ts (hunk 8 above)
Comment (m-ujaffer) [reply src/services/TechRecordsService.ts]: fixed

File: tests/unit/handler.unitTest.ts (hunk 9 above)
Comment (m-ujaffer) [reply tests/unit/handler.unitTest.ts]: used mockName.mock.calls[0][0] and mockName.mock.calls[0][1]

File: src/config/config.yml (hunk 7 above)
Comment (m-ujaffer) [reply src/config/config.yml]: fixed

File: src/assets/Enums.ts (hunk 6 above)
Comment (m-ujaffer) [reply src/assets/Enums.ts]: fixed

File: tests/integration/updateEuVehicleCategroy.intTest.ts [hunk 10]
```diff
@@ -61,7 +61,7 @@ describe("updateEuVehicleCategory", () => {
                     expect(result.statusCode).toBe(200);
                     const techRecordWrapper: ITechRecordWrapper = JSON.parse(result.body);
                     expect(techRecordWrapper.techRecord.length).toBe(2);
-                    expect(techRecordWrapper.techRecord[0].euVehicleCategory).toBe(null);
+                    expect(techRecordWrapper.techRecord[0].euVehicleCategory).toBe(undefined);
```
Comment (agoncearenco) [inline tests/integration/updateEuVehicleCategroy.intTest.ts]: the logic checks `if(nonArchivedTechRecord[0].euVehicleCategory)`. Shouldn't you have tests for both null and undefined?

Comment (m-ujaffer) [reply tests/integration/updateEuVehicleCategroy.intTest.ts]: added a test for null value as well

**Pull request merged**